src/ORM/DataList.php

```diff
@@ -308,72 +313,106 @@ public function distinct($value)
+            if (is_string($sort)) {
+                $newSort = [];
+                // Making the assumption here there are no commas in column names
+                // Other parts of silverstripe will break if there are commas in column names
+                foreach (explode(',', $sort) as $colDir) {
+                    if (empty($colDir)) {
+                        continue;
+                    }
+                    // Using regex instead of explode(' ') in case column name includes spaces
+                    if (preg_match('/^(.+) (asc|desc)$/i', trim($colDir), $matches)) {
+                        list($column, $direction) = [$matches[1], $matches[2]];
+                    } else {
+                        list($column, $direction) = [$colDir, 'ASC'];
+                    }
+                    $newSort[$column] = $direction;
+                }
+                $sort = $newSort;
+            }
+        }
+        // Validate that column names are not raw sql
+        if (is_array($sort)) {
+            foreach (array_keys($sort) as $column) {
+                $this->validateColumnHasNoRawSql($column);
+            }
         }
-
         return $this->alterDataQuery(function (DataQuery $query, DataList $list) use ($sort) {
-
             if (is_string($sort) && $sort) {
-                if (false !== stripos($sort ?? '', ' asc') || false !== stripos($sort ?? '', ' desc')) {
+                if (preg_match('/^(.+) (asc|desc)$/i', $sort)) {
                     $query->sort($sort);
                 } else {
                     $list->applyRelation($sort, $column, true);
                     $query->sort($column, 'ASC');
                 }
             } elseif (is_array($sort)) {
-                // sort(array('Name'=>'desc'));
-                $query->sort(null, null); // wipe the sort
-
+                // Wipe the sort
+                $query->sort(null, null);
                 foreach ($sort as $column => $direction) {
-                    // Convert column expressions to SQL fragment, while still allowing the passing of raw SQL
-                    // fragments.
                     $list->applyRelation($column, $relationColumn, true);
                     $query->sort($relationColumn, $direction, false);
                 }
             }
         });
     }
 
+    private function validateColumnHasNoRawSql(string $column): void
+    {
+        // Remove quotes around the column name
+        $col = trim(str_replace('"', '', $column));
+        // Useing regex because $this->canSortBy() failed on complex relations e.g. Team.Founder.Surname and
+        // also increased the number of db queries
+        // Search for non-alphanumeric characters in the column name rather than ^[^a-zA-Z0-9_]$ to account
+        // for column names that include macrons etc
+        if (preg_match('#[\(\)\[\]{}\+\-\*/=\:\?\\;|&\'"`<>^%, ]#', $col)) {
+            throw new InvalidArgumentException("Invalid column name $column");
```

Comment:
I've updated the message.

I think enforcing column names is beyond the scope of this issue, it needs thought because there could be some legacy projects where for whatever reason there are strange column names with dashes in the column names, but they cannot easily be renamed because of systems may rely on those column names.

In the message I've suggested people either rename the column or use orderBy() instead

Reply:
Dashes and underscores seem like the most likely non-alpha-numerical characters that could be in column names, and it seems unlikely to me that they could be used to cause an SQL Injection vulnerability. I think we should remove hyphens from this regex (underscores are already not in it)

Reply:
Also, FYI, I've created https://github.com/silverstripe/silverstripe-framework/issues/10605 for the wider concern.

Reply:
Have refactored out the regex and now using existing API on DataList to validate